manager/integration/tests/test_scheduling.py

```diff
@@ -0,0 +1,349 @@
+    client.update(setting, value="false")
+    node = client.by_id_node(host_id)
+    client.update(node, allowScheduling=False)
+    replica_names = map(lambda replica: replica.name, volume["replicas"])
+    host_replica = get_host_replica(volume, host_id)
+
+    volume.replicaRemove(name=host_replica["name"])
+    wait_new_replica_ready(client, volume_name, replica_names)
+    volume = wait_for_volume_healthy(client, volume_name)
+    assert len(volume["replicas"]) == 3
+    check_volume_data(volume, data)
+
+    cleanup(client, volume)
+
+
+def test_soft_anti_affinity_detach(client, volume_name):  # NOQA
+    """
+    Test that volumes with Soft Anti-Affinity can detach and reattach to a
+    node properly.
+    """
+    volume = create_and_check_volume(client, volume_name)
+    host_id = get_self_host_id()
+    volume.attach(hostId=host_id)
+    volume = wait_for_volume_healthy(client, volume_name)
+    assert len(volume["replicas"]) == 3
+
+    data = write_volume_random_data(volume)
+    setting = client.by_id_setting(SETTING_REPLICA_HARD_ANTI_AFFINITY)
+    client.update(setting, value="false")
+    node = client.by_id_node(host_id)
+    client.update(node, allowScheduling=False)
+    replica_names = map(lambda replica: replica.name, volume["replicas"])
+    host_replica = get_host_replica(volume, host_id)
+
+    volume.replicaRemove(name=host_replica["name"])
+    wait_new_replica_ready(client, volume_name, replica_names)
+    volume = wait_for_volume_healthy(client, volume_name)
+    volume.detach()
+    volume = wait_for_volume_detached(client, volume_name)
+    assert len(volume["replicas"]) == 3
+
+    volume.attach(hostId=host_id)
+    volume = wait_for_volume_healthy(client, volume_name)
+    assert len(volume["replicas"]) == 3
+    check_volume_data(volume, data)
+
+    cleanup(client, volume)
+
+
+def test_hard_anti_affinity_scheduling(client, volume_name):  # NOQA
+    """
+    Test that volumes with Hard Anti-Affinity work as expected.
+
+    With Hard Anti-Affinity, scheduling on nodes with existing replicas should
+    be forbidden, resulting in "Degraded" state.
+    """
+    volume = create_and_check_volume(client, volume_name)
+    host_id = get_self_host_id()
+    volume.attach(hostId=host_id)
+    volume = wait_for_volume_healthy(client, volume_name)
+    assert len(volume["replicas"]) == 3
+
+    data = write_volume_random_data(volume)
+    setting = client.by_id_setting(SETTING_REPLICA_HARD_ANTI_AFFINITY)
+    client.update(setting, value="true")
+    node = client.by_id_node(host_id)
+    client.update(node, allowScheduling=False)
+    replica_names = map(lambda replica: replica.name, volume["replicas"])
+    host_replica = get_host_replica(volume, host_id)
+
+    volume.replicaRemove(name=host_replica["name"])
+    wait_no_new_replica_ready(client, volume_name, replica_names)
+    volume = wait_for_volume_degraded(client, volume_name)
+    # We can't just count the replicas since there may be more than two due to
```

Comment:
You've seen the manager attempted to schedule more than one replicas? That doesn't sound right.

Reply:
Since the system is attempting to schedule a new `Replica`, that `Replica` will still appear in the list of replicas. Even though the scheduling fails and the `Replica` gets deleted, there's still the chance that if we check the number of replicas, we'll pick up the third one that the `longhorn-manager` is attempting to schedule, even though that one will eventually fail and delete.

The most I've seen is a fourth `Replica` during this rescheduling process, which should be correct because the third `Replica` is failed due to not being scheduled, allowing for the `longhorn-manager` to attempt to schedule a fourth one.